manager/allocator/network/ipam/errors.go

```diff
@@ -0,0 +1,201 @@
+}
+
+// IsErrBustedIPAM returns true if the type of the error is ErrBustedIPAM
+func IsErrBustedIPAM(e error) bool {
+	_, ok := e.(ErrBustedIPAM)
+	return ok
+}
+
+// ErrInvalidAddress is an error type indicating that a requested address's
+// string form is not a valid address and it cannot be parsed.
+type ErrInvalidAddress struct {
+	address string
+}
+
+// Error returns a formatted error message explaining which address is invalid
+func (e ErrInvalidAddress) Error() string {
+	return fmt.Sprintf("address %v is not a valid IP address", e.address)
+}
+
+// IsErrInvalidAddress returns true if the type of the error is
+// ErrInvalidAddress.
+func IsErrInvalidAddress(e error) bool {
+	_, ok := e.(ErrInvalidAddress)
+	return ok
+}
+
+// ErrBadState is the error type returned when restoring the state of the
+// cluster to the local state encounters a problem that is likely a result of
+// inconsistent state. It includes the network id and a description of the
+// problem
+type ErrBadState struct {
+	nwid    string
+	problem string
+}
+
+// Error returns a formatted error string explaining what went wrong and with
+// what network
+func (e ErrBadState) Error() string {
+	return fmt.Sprintf("restoring ipam for network %v encountered a bad state: %v", e.nwid, e.problem)
+}
+
+// IsErrBadState returns true if the type of the error is ErrBadState
+func IsErrBadState(e error) bool {
+	_, ok := e.(ErrBadState)
+	return ok
+}
+
+// ErrNetworkAllocated is an error type returned when a user tries to call
+// AllocateNetwork with a network that has already been allocated, because
+// network updates are not supported.
+type ErrNetworkAllocated struct {
+	nwid string
+}
+
+// Error returns a formatted error message explaining that the network is
+// already allocated and cannot be updated
+func (e ErrNetworkAllocated) Error() string {
+	return fmt.Sprintf(
+		"network %v is already allocated and network updates are not supported",
+		e.nwid,
+	)
+}
+
+// IsErrNetworkAllocated returns true if the type of the error is
+// ErrNetworkAllocated
+func IsErrNetworkAllocated(e error) bool {
+	_, ok := e.(ErrNetworkAllocated)
+	return ok
+}
+
+// ErrFailedPoolRequest is returned if requesting an IPAM pool fails
+type ErrFailedPoolRequest struct {
+	subnet, iprange string
+	err             error
+}
+
+// Error prints an error message explaining why the pool request failed
+func (e ErrFailedPoolRequest) Error() string {
+	return fmt.Sprintf(
+		"requesting pool (subnet: %q, range: %q) returned error: %v",
+		e.subnet, e.iprange, e.err,
+	)
+}
+
+// IsErrFailedPoolRequest returns true if the type of the error is ErrInvalidPool
+func IsErrFailedPoolRequest(e error) bool {
+	_, ok := e.(ErrFailedPoolRequest)
+	return ok
+}
+
+// ErrFailedAddressRequest is the error type used if an address request to the
+// IPAM, either for a gateway, an endpoint, or an attachment, fails for any
+// reason.
+type ErrFailedAddressRequest struct {
+	address string
+	err     error
```

Comment:
We certainly can do that. On a local branch I've restructured the errors so that the same set of errors is shared across all allocator subcomponents. I'll push up to this branch what progress I've made on the IPAM allocator tomorrow and we can take another look.